File: frontend/src/services/productService.js
import api from './api';

export const productService = {
    getAllProducts: async (filters = {}) => {
        try {
            let url = '/products';
            const params = {};

            // Handle sorting
            if (filters.sort) {
                if (filters.sort === 'price_asc') {
                    params.sortBy = 'price';
                    params.direction = 'ASC';
                } else if (filters.sort === 'price_desc') {
                    params.sortBy = 'price';
                    params.direction = 'DESC';
                } else if (filters.sort === 'rating') {
                    params.sortBy = 'rating';
                    params.direction = 'DESC';
                }
            }

            // Handle search
            if (filters.search) {
                url = '/products/search';
                params.keyword = filters.search;
            }
            // Handle category
            else if (filters.category) {
                // Assuming filters.category is the ID. If it's a name, we might need a different approach or backend change.
                // For now, let's try to use the category endpoint if we have an ID.
                // If the frontend passes a name, this will fail unless we map it.
                // However, standard practice is to use IDs. 
                // If the UI passes a name, we might need to fetch all categories to find the ID, but that's inefficient.
                // Let's assume we can pass the ID.
                url = `/products/category/${filters.category}`;
            }

            const response = await api.get(url, { params });
            return response.data;
        } catch (error) {
            console.error('Error fetching products:', error);
            throw error;
        }
    },

    getProductById: async (id) => {
        try {
            const response = await api.get(`/products/${id}`);
            return response.data;
        } catch (error) {
            console.error(`Error fetching product ${id}:`, error);
            throw error;
        }
    },

    getCategories: async () => {
        try {
            const response = await api.get('/categories');
            return response.data; // Returns List<Category> objects {id, name, ...}
        } catch (error) {
            console.error('Error fetching categories:', error);
            throw error;
        }
    },

    getRecommendations: async (customerId) => {
        try {
            if (!customerId) return [];
            const response = await api.get(`/recommendations/${customerId}`);
            return response.data;
        } catch (error) {
            console.error('Error fetching recommendations:', error);
            return [];
        }
    }
};

File: api/src/main/java/com/mycompany/ecommerce/api/ProductService.java
package com.mycompany.ecommerce.api;

import com.mycompany.ecommerce.api.model.Category;
import com.mycompany.ecommerce.api.model.Product;

import java.util.List;

public interface ProductService {

    // UC-CS-01 Search Products
    List<Product> searchProducts(String keyword);

    List<Product> getAllProducts();

    // UC-CS-02 Browse Categories
    List<Category> getAllCategories();

    List<Product> getProductsByCategory(Long categoryId);

    // UC-CS-03 View Recommendations
    List<Product> getRecommendations(Long customerId);

    // UC-CS-04 Sort Results
    // Sorting can be a parameter in search or a separate helper,
    // but usually handled by passing a sort param to search/category methods.
    // We will overload search/category methods or just assume list processing in
    // implementation/controller if data set is small.
    // For now, let's keep it simple.

    Product getProduct(Long id);
}
